core/src/main/java/org/apache/druid/math/expr/ApplyFunction.java

```diff
@@ -0,0 +1,611 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.math.expr;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import it.unimi.dsi.fastutil.objects.Object2IntArrayMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.druid.java.util.common.RE;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+public interface ApplyFunction
+{
+  String name();
+
+  ExprEval apply(LambdaExpr lambdaExpr, List<Expr> argsExpr, Expr.ObjectBinding bindings);
+
+  Set<Expr> getArrayInputs(List<Expr> args);
+
+  abstract class BaseMapFunction implements ApplyFunction
+  {
+    ExprEval applyMap(LambdaExpr expr, int length, IndexableMapLambdaObjectBinding bindings)
+    {
+      String[] stringsOut = null;
+      Long[] longsOut = null;
+      Double[] doublesOut = null;
+
+      ExprType outputType = null;
+      Object out = null;
+      for (int i = 0; i < length; i++) {
+
+        ExprEval evaluated = expr.eval(bindings.withIndex(i));
+        if (outputType == null) {
+          outputType = evaluated.type();
+          switch (outputType) {
+            case STRING:
+              stringsOut = new String[length];
+              out = stringsOut;
+              break;
+            case LONG:
+              longsOut = new Long[length];
+              out = longsOut;
+              break;
+            case DOUBLE:
+              doublesOut = new Double[length];
+              out = doublesOut;
+              break;
+            default:
+              throw new RE("Unhandled map function output type [%s]", outputType);
+          }
+        }
+
+        switch (outputType) {
+          case STRING:
+            stringsOut[i] = evaluated.asString();
+            break;
+          case LONG:
+            longsOut[i] = evaluated.asLong();
+            break;
+          case DOUBLE:
+            doublesOut[i] = evaluated.asDouble();
+            break;
+        }
+      }
+      return ExprEval.bestEffortOf(out);
```

Comment:
when we know the type of result then why do best effort ?

Reply:
Good point, will fix :+1: